bundles/ui/org.eclipse.smarthome.ui.icon/src/main/java/org/eclipse/smarthome/ui/icon/internal/IconServlet.java

```diff
@@ -97,21 +102,62 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
             return;
         }
 
-        String filename = StringUtils.substringAfterLast(req.getRequestURI(), "/");
-
-        String iconName = StringUtils.substringBeforeLast(filename, ".");
+        String category = getCategory(req);
+        Format format = getFormat(req);
+        String state = getState(req);
+        String iconSetId = getIconSetId(req);
 
+        IconProvider topProvider = null;
+        int maxPrio = Integer.MIN_VALUE;
         for (IconProvider provider : iconProvider) {
-            if (provider.hasIcon(iconName)) {
+            Integer prio = provider.hasIcon(category, iconSetId, format);
+            if (prio != null && prio > maxPrio) {
+                maxPrio = prio;
+                topProvider = provider;
+            }
+        }
+        if (topProvider != null) {
+            if (format.equals(Format.SVG)) {
+                resp.setContentType("image/svg+xml");
+            } else {
                 resp.setContentType("image/png");
-                resp.setDateHeader("Last-Modified", new Date().getTime());
-                ServletOutputStream os = resp.getOutputStream();
-                InputStream is = provider.getIcon(iconName);
-                IOUtils.copy(is, os);
-                resp.flushBuffer();
-                return;
             }
+            resp.setDateHeader("Last-Modified", new Date().getTime());
+            ServletOutputStream os = resp.getOutputStream();
+            InputStream is = topProvider.getIcon(category, iconSetId, state, format);
+            IOUtils.copy(is, os);
+            is.close();
+            resp.flushBuffer();
+        } else {
+            resp.sendError(404);
         }
-        resp.sendError(404);
+    }
+
+    private String getCategory(HttpServletRequest req) {
+        String filename = StringUtils.substringAfterLast(req.getRequestURI(), "/");
+        return StringUtils.substringBeforeLast(filename, ".");
```

Comment:
This is for backward compatibility, so that clients that still send a /icon/switch.png still work.